c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,350 @@
+    "must be `FLOAT64`.\n"
+    "NB: as the model trains, this frame will be changed in-place.\n");
+
+  gs.add<&Ftrl::get_a, &Ftrl::set_a>("a", "`alpha` in per-coordinate learning rate formula.\n");
+  gs.add<&Ftrl::get_b, &Ftrl::set_b>("b", "`beta` in per-coordinate learning rate formula.\n");
+  gs.add<&Ftrl::get_l1, &Ftrl::set_l1>("l1", "L1 regularization parameter.\n");
+  gs.add<&Ftrl::get_l2, &Ftrl::set_l2>("l2", "L2 regularization parameter.\n");
+  gs.add<&Ftrl::get_d, &Ftrl::set_d>("d", "Number of bins to be used after the hashing trick.\n");
+  gs.add<&Ftrl::get_n_epochs, &Ftrl::set_n_epochs>("n_epochs", "Number of epochs to train for.\n");
+  gs.add<&Ftrl::get_inter, &Ftrl::set_inter>("inter", "If feature interactions to be used or not.\n");
+  gs.add<&Ftrl::get_hash_type, &Ftrl::set_hash_type>("hash_type", "Hashing method to use for strings.\n"
+    "`0` - std::hash;\n"
+    "`1` - Murmur2;\n"
+    "`2` - Murmur3.\n");
+  gs.add<&Ftrl::get_seed, &Ftrl::set_seed>("seed", "Seed to be used for Murmur hash functions.\n");
+
+  mm.add<&Ftrl::fit, args_fit>();
+  mm.add<&Ftrl::predict, args_predict>();
+}
+
+
+PKArgs Ftrl::Type::args_fit(1, 0, 0, false, false, {"frame"}, "fit",
+R"(fit(self, frame)
+--
+
+Train an FTRL model on a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame to be trained on, last column is treated as `target`.
+
+Returns
+----------
+    None
+)");
+
+
+void Ftrl::fit(const PKArgs& args) {
+  DataTable* dt_train = args[0].to_frame();
+  fm->fit(dt_train);
+}
+
+
+PKArgs Ftrl::Type::args_predict(1, 0, 0, false, false, {"frame"}, "predict",
+R"(predict(self, frame)
+--
+
+Make predictions for a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame of shape `(nrows, ncols)` to make predictions for. It must have one 
+    column less than the training dataset.
+
+Returns
+----------
+    A new `Frame` of shape `(nrows, 1)` with a prediction for each row.
+)");
+
+
+oobj Ftrl::predict(const PKArgs& args) {
+  DataTable* dt_test = args[0].to_frame();
+  DataTable* dt_target = fm->predict(dt_test).release();
+  py::oobj df_target = py::oobj::from_new_reference(py::Frame::from_datatable(dt_target));
+
+  return df_target;
+}
+
+
+/*
+*  Getter and setter for the model datatable.
+*/
+oobj Ftrl::get_model(void) const {
+  DataTable* dt_model = fm->get_model();
+  py::oobj df_model = py::oobj::from_new_reference(py::Frame::from_datatable(dt_model));
+  return df_model;
+}
+
+
+void Ftrl::set_model(robj model) {
+  DataTable* dt_model_in = model.to_frame();
+  const std::vector<std::string>& model_cols_in = dt_model_in->get_names();
+
+  if (dt_model_in->nrows != fm->get_d() || dt_model_in->ncols != 2) {
+    throw ValueError() << "FTRL model frame must have " << fm->get_d() << " rows,"
+                       << "and 2 columns, whereas your frame has " << dt_model_in->nrows
+                       << " rows and " << dt_model_in->ncols << " columns";
+  } else if (model_cols_in != FtrlModel::model_cols) {
```

Comment:
No need to say `else if` if the previous `if()` condition throws an error or returns.
The code will look cleaner if you write
```
if (dt_model_in->nrows != fm->get_d() || dt_model_in->ncols != 2) { ... }
if (model_cols_in != FtrlModel::model_cols) { ... }
if (dt_model_in->columns[0]->stype() != SType::FLOAT64 ||
    dt_model_in->columns[1]->stype() != SType::FLOAT64) { ... }

fm->set_model(dt_model_in);
```

Reply:
Sure, I'm still looking through the code and making some adjustments.